### PR TITLE
Add support for pnpm

### PR DESCRIPTION
**Pull request by cd-work**
This adds support for the new `pnpm-lock.yaml` lockfile type and its
package manager `pnpm`. Supported are both parsing the lockfile itself
and generating a lockfile from the `package.json`.

Closes https://github.com/phylum-dev/cli/issues/1138.

### CODE REVIEW

File: lockfile/src/javascript.rs [hunk 1]
```diff
@@ -263,6 +267,69 @@ impl Parse for YarnLock {
...
+
+impl PnpmLock {
+    /// Get all packages in the lockfile.
+    fn packages(&self) -> anyhow::Result<Vec<Package>> {
+        let mut packages = Vec::new();
+
+        // Convert `/@org/package@1.2.3(test@3.2.1)` to package.
+        for name in self.packages.keys() {
+            // Strip `/` prefix.
+            let name = &name[1..];
```
Comment (kylewillmon) [inline lockfile/src/javascript.rs:311]: It's probably safer to confirm that `name[0]` is a `/` before stripping it. Something like:

```rust
let name = name.strip_prefix("/").ok_or_else(|| anyhow!("Invalid package: {name}"))?;
```

**Review by kylewillmon — CHANGES_REQUESTED** [review]
I know that `pnpm` (like `yarn`) supports [patching dependencies](https://pnpm.io/feature-comparison).

I assume they also support filesystem dependencies and VCS dependencies. But the current code doesn't seem to handle any of that. Do those not show up in the `packages` list?

**Review by maxrake — DISMISSED** [review]
The code was reviewed and looked okay to my eye. The changes were tested locally with a few real-world `pnpm-lock.yaml` files and worked as expected.

The requested change is to add a change log entry.

**Comment by cd-work** [comment]
> I assume they also support filesystem dependencies and VCS dependencies.

I tested this and filesystem dependencies indeed do not show up in the packages list.

VCS dependencies I didn't take into account and they do indeed seem to be supported with a slightly different syntax.

**Comment by cd-work** [comment]
I've added support for tarballs. Neither git nor path dependencies seem to show up in the `packages` section of the lockfile. They do show up at the top as dependencies, but I don't think it's worth trying to parse it considering we don't analyze them anyway.

**Comment by cd-work** [comment]
Patched dependencies show up as their original dependency with some minor patch information associated. I think it makes sense to just treat them as firstparty dependencies since that we can actually analyze.

**Review by kylewillmon — APPROVED** [review]

**Pull request merged**